posthog/api/plugin.py

```diff
@@ -0,0 +1,132 @@
+import json
+from typing import Any, Dict
+
+import requests
+from django.conf import settings
+from rest_framework import request, serializers, viewsets
+from rest_framework.decorators import action
+from rest_framework.exceptions import APIException
+from rest_framework.response import Response
+
+from posthog.models import Plugin, PluginConfig
+from posthog.plugins import download_plugin_github_zip, reload_plugins_on_workers
+
+
+class PluginSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Plugin
+        fields = ["id", "name", "description", "url", "config_schema", "tag", "from_json"]
+
+    def create(self, validated_data: Dict, *args: Any, **kwargs: Any) -> Plugin:
+        if not settings.PLUGINS_INSTALL_FROM_WEB:
+            raise APIException("Plugin installation via the web is disabled!")
+        if len(Plugin.objects.filter(name=validated_data["name"])) > 0:
+            raise APIException('Plugin with name "{}" already installed!'.format(validated_data["name"]))
+        validated_data["archive"] = download_plugin_github_zip(validated_data["url"], validated_data["tag"])
+        if "from_json" in validated_data:  # prevent hackery
+            del validated_data["from_json"]
+        plugin = Plugin.objects.create(from_web=True, **validated_data)
+        reload_plugins_on_workers()
+        return plugin
+
+    def update(self, plugin: Plugin, validated_data: Dict, *args: Any, **kwargs: Any) -> Plugin:  # type: ignore
+        if not settings.PLUGINS_INSTALL_FROM_WEB:
+            raise APIException("Plugin installation via the web is disabled!")
+        if plugin.from_json:
+            raise APIException('Can not update plugin "{}", which is configured from posthog.json!'.format(plugin.name))
+        plugin.name = validated_data.get("name", plugin.name)
+        plugin.description = validated_data.get("description", plugin.description)
+        plugin.url = validated_data.get("url", plugin.url)
+        plugin.config_schema = validated_data.get("config_schema", plugin.config_schema)
+        plugin.tag = validated_data.get("tag", plugin.tag)
+        plugin.archive = download_plugin_github_zip(plugin.url, plugin.tag)
+        plugin.save()
+        reload_plugins_on_workers()
+        return plugin
+
+
+class PluginViewSet(viewsets.ModelViewSet):
+    queryset = Plugin.objects.all()
+    serializer_class = PluginSerializer
+
+    def get_queryset(self):
+        queryset = super().get_queryset()
+        if not settings.PLUGINS_INSTALL_FROM_WEB and not settings.PLUGINS_CONFIGURE_FROM_WEB:
```

Comment:
Is `PLUGINS_CONFIGURE_FROM_WEB` intended not to allow even to list which plugins are installed? Genuine question

Reply:
You should be able to list what plugins are installed, as if you only have `PLUGINS_CONFIGURE_FROM_WEB` and not `PLUGINS_INSTALL_FROM_WEB` (e.g. like it would probably be for `app.posthog.com`), you should see the plugins and configure/enable them as needed.

The if here might have been confusing. I changed it from:

```py
        if not settings.PLUGINS_INSTALL_FROM_WEB and not settings.PLUGINS_CONFIGURE_FROM_WEB:
            return queryset.none()
        return queryset
```

to:

```py
        if settings.PLUGINS_INSTALL_FROM_WEB or settings.PLUGINS_CONFIGURE_FROM_WEB:
            return queryset
        return queryset.none()
```